### PR TITLE
Add a futures universe selection model that uses open interest

### PR DESCRIPTION
**Pull request by michael-sena**
Adds a purposed futures universe selection model, that will assist with identifying the current active contract / contracts.

#### Description
Universe selection is done by obtaining the last open interest for each contract.  The contracts are then sorted by this value, and provided back for use.

#### Related Issue
N/A

#### Motivation and Context
While the front month in a futures chain is often also the most outstanding orders, this isn't always the case. In order to select the right contract in the chain, open interest should be used to sort the contracts.

This model provides an easy way for that to be done.

#### Requires Documentation Change
N/A

#### How Has This Been Tested?
Yes using a broker, and comparing the broker's expected active contract to the returned results.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>` - *it is missing the issue, does that matter?*

Credits also to @StefanoRaggi for his assistance in this model, and the original version of this class.

### CODE REVIEW

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using QuantConnect.Algorithm.Framework.Selection;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Algorithm.Farmework.Selection
```
Comment (StefanoRaggi) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:27]: Small typo here

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        /// <returns>Filtered set</returns>
+        public IEnumerable<Symbol> FilterByOpenInterest(IEnumerable<Symbol> contracts)
+        {
+            return contracts.OrderBy(x => x.ID.Date).Take(_chainContractsLookupLimit).OrderByDescending(GetOpenInterest).ThenBy(x => x.ID.Date).Take(_resultsLimit);
+        }
+
+        private decimal GetOpenInterest(Symbol symbol)
+        {
+            var current = _algorithm.Time;
+            var previousBusinessDate = GetPreviousBusinessDay(_algorithm.Time, symbol);
```
Comment (StefanoRaggi) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:80]: Doesn't this assume the algorithm time zone defaulting to `TimeZones.NewYork`? 
Maybe it's better to use `algorithm.UtcTime` instead?

File: Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+using QuantConnect.Configuration;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Tests.Algorithm.Framework.Selection
+{
+    [TestFixture]
+    public class OpenInterestFutureUniverseSelectionModelTests
```
Comment (StefanoRaggi) [inline Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs:18]: Nice tests! 👍

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                            null,
+                            true,
+                            false,
+                            DataNormalizationMode.Raw,
+                            TickType.OpenInterest
+                        )
+                    },
+                    TimeZones.Utc
+                )
+                .FirstOrDefault();
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:101]: Maybe we should get the `lastOrDefault()` instead?

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            {
+                var message = new StringBuilder(nameof(OpenInterestFutureUniverseSelectionModel)).Append('.')
+                    .Append(MethodBase.GetCurrentMethod().Name)
+                    .Append("(): Received ")
+                    .Append(ticks.Count)
+                    .Append(" entries for ")
+                    .Append(symbol.ToString())
+                    .Append(" tick data containing open interest. Will use first. Values: ")
+                    .Append(string.Join(",", ticks));
+                _algorithm.Error(message.ToString());
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:120]: Not sure this is an error? The history call could return more than 1 data point using tick resolution

File: Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        };
+        private Mock<IHistoryProvider> _mockHistoryProvider;
+        private OpenInterestFutureUniverseSelectionModel _underTest;
+
+        [SetUp]
+        public void SetUp()
+        {
+            Config.Reset();
+            MarketHoursDatabase.Reset();
+            Config.Set("data-folder", Path.Combine(TestContext.CurrentContext.TestDirectory, "..", "..", "..", "Data"));
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs:43]: Wondering why are these required, I gave it a quick test and it seemed to work correctly without them too?
```
            Config.Reset();
            MarketHoursDatabase.Reset();
            Config.Set("data-folder", Path.Combine(TestContext.CurrentContext.TestDirectory, "..", "..", "..", "Data"));
```

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            }
+
+            _algorithm = algorithm;
+            _resultsLimit = resultsLimit;
+            _chainContractsLookupLimit = chainContractsLookupLimit;
+            _exchangeHours = SecurityExchangeHours.AlwaysOpen(TimeZones.Utc);
+        }
+
+        /// <inheritdoc cref="FutureUniverseSelectionModel.Filter" />
+        protected override FutureFilterUniverse Filter(FutureFilterUniverse filter)
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:62]: > /// inheritdoc cref="FutureUniverseSelectionModel.Filter" 

I believe we don't use `inheritdoc` because it broke the documentation generator

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                        )
+                    },
+                    TimeZones.Utc
+                )
+                .FirstOrDefault();
+
+            var ticks = result?.Ticks?[symbol];
+            if (ticks == null || ticks.Count == 0)
+            {
+                _algorithm.Log($"WARN: Unable to retrieve open interest for: {symbol}");
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: Does this happen often? I think we should add "OpenInterestFutureUniverseSelectionModel" in the log message

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Algorithm.Farmework.Selection
+{
+    /// <summary>
+    ///     Selects contracts in a futures universe, sorted by open interest.  This allows the selection to identifiy current
+    ///     active contract.
+    /// </summary>
+    public class OpenInterestFutureUniverseSelectionModel : FutureUniverseSelectionModel
```
[the remark this answers is not in the document]
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:33]: What do you mean by that sorry?

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:33]: Like this guy `BasicTemplateFuturesFrameworkAlgorithm` https://github.com/QuantConnect/Lean/blob/master/Algorithm.CSharp/BasicTemplateFuturesFrameworkAlgorithm.cs, a small simple algorithm which uses open-source available data to assert the behavior of the selection model and is an example of it usage.
*No need to define a custom alpha or PortfolioConstructionModel as `BasicTemplateFuturesFrameworkAlgorithm` does, just using the new `OpenInterestFutureUniverseSelectionModel` and placing at least 1 trade (as a way of asserting the behavior)

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 7 above)
Comment (jaredbroad) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:62]: Because we try to make all our comments readable without navigating so it is easier to understand in a web environment like Github/QC

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 5 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:120]: Removed

File: Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs (hunk 6 above)
Comment (michael-sena) [reply Tests/Algorithm/Framework/Selection/OpenInterestFutureUniverseSelectionModelTests.cs:43]: Left-over from code in our own project. :P

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 8 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: The contracts being provided in should all be active, so not having a value doesn't sound right to me.  If it were me, I'd rather it be an error - but I figured others would probably only want a warning.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 7 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:62]: I've added the full doc.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 8 above)
Comment (RichardDale) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: Note that most futures exchanges report open interest significantly after that trading session has finished, so it's quite common for it to be null (unknown) until that time.

Other exchanges provide an estimate then update it a day later.

It's certainly possible to pad it forward but it would be stale versus the other data points.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 4 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:101]: Changed to taking the last, should there be multiple values.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            var previousBusinessDate = GetPreviousBusinessDay(_algorithm.Time, symbol);
+            var result = _algorithm.HistoryProvider.GetHistory(
+                    new[]
+                    {
+                        new HistoryRequest(
+                            previousBusinessDate,
+                            current,
+                            typeof(Tick),
+                            symbol,
+                            Resolution.Tick,
```
[the remark this answers is not in the document]
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:89]: I've changed to using the exchange hours as suggested.  Looks good. 

As for Resolution.Minute, for Open Interest, my understanding is that it only updates daily.  So the intent is to only get the most recent value, quickly.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+        /// <summary>
+        ///     Creates a new instance of <see cref="OpenInterestFutureUniverseSelectionModel" />
+        /// </summary>
+        /// <param name="algorithm">Algorithm</param>
+        /// <param name="futureChainSymbolSelector">Selects symbols from the provided future chain</param>
+        /// <param name="chainContractsLookupLimit">Limit on how many contracts to query for open interest</param>
+        /// <param name="resultsLimit">Limit on how many contracts will be part of the universe</param>
+        public OpenInterestFutureUniverseSelectionModel(IAlgorithm algorithm, Func<DateTime, IEnumerable<Symbol>> futureChainSymbolSelector, int chainContractsLookupLimit = 6,
+            int resultsLimit = 1) : base(TimeSpan.FromDays(1), futureChainSymbolSelector)
```
[the remark this answers is not in the document]
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:48]: Added

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 2 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:80]: Addressed, as mentioned in lower conversations.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 1 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:27]: Fixed

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:33]: Okay, I've added a basic regression algorithm.  It really doesn't do much other than have in the statistics the number of expected trades be 4 - which matches the limit on the symbols.

I'm not sure how beneficial it is though, because it seems like the regression suite doesn't use actual data - and therefore, the historical look-up of open interest isn't providing anything of significance.

Do you have any suggestions?

Otherwise, is it worth keeping?

**Comment by michael-sena** [comment]
@Martin-Molinero - With the batch history lookups, while history look-ups are passed to the provider in bulk, I believe if it's then passed to a broker, the broker can only receive a single requests?  (Looking at: `IBrokerage.GetHistory(HistoryRequest request)`.  May be worth allowing brokers (if they're able to) to query in bulk.

**Comment by michael-sena** [comment]
I believe I've addressed all comments now, and resolved conversations outstanding with me.  Awaiting your review!

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 8 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: Let me know if you're happy with the approach I've taken @RichardDale, following @Martin-Molinero's suggestions.

Comment (RichardDale) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: There are implications for an open interest data point update changing the Universe "after the fact" especially where you have futures rolls occurring on the basis of open interest.  

Most futures traders understand this already but it's certainly a caveat emptor if you're trading spot month crude oil futures!

Otherwise, the approach looks good.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:89]: Cool, your right! Sounds good 👍

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:33]: >  It really doesn't do much other than have in the statistics the number of expected trades

That's okay 👍 my suggestion was adding it as a simple example and as a regression test (which asserts on the stats)

> doesn't use actual data

How/on what is it trading? Open-source data folder has GC future tick data available with open interest https://github.com/QuantConnect/Lean/tree/master/Data/future/comex/tick/gc, using the GC future in those available days should work for using the `OpenInterestFutureUniverseSelectionModel` and placing some trade?

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 8 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:106]: As far as I know, there isn't any mechanism to do data revisions in Quant Connect.  So I this aspect is a bit baked into the cake, but I'll leave it up to @Martin-Molinero to decide if this can be resolved or addressed.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:33]: Okay, I've updated it again with an additional check.  Hopefully it's a bit more inline with the expected approach now.

It's trading gold, but it's test is to look at the contract selection.

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:32]: Looks great ty 👍

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 8 above)
Comment (Martin-Molinero) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs]: > there isn't any mechanism to do data revisions in Quant Connect

Raw _backtesting_ data is available at https://www.quantconnect.com/data/tree
Regarding live trading, I believe it will drill down to the history provider being used. The history request is at the most 1 day backward with no fill forwarding from which we take the last available value. If desired, the backward history request window could be reduced, but as is seems good to me 👍

File: Algorithm.CSharp/OpenInterestFuturesRegressionAlgorithm.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        /// <summary>
+        ///     This is used by the regression test system to indicate if the open source Lean repository has the required data to
+        ///     run this algorithm.
+        /// </summary>
+        public bool CanRunLocally { get; } = true;
+
+        /// <summary>
+        ///     This is used by the regression test system to indicate which languages this algorithm is written in.
+        /// </summary>
+        public Language[] Languages { get; } = {Language.CSharp, Language.Python};
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/OpenInterestFuturesRegressionAlgorithm.cs:90]: nit - `Language.Python` should be removed, else it will try loading the unexisting python version of this algorithm.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        /// </summary>
+        /// <param name="contracts">Contracts to filter</param>
+        /// <returns>Filtered set</returns>
+        public IEnumerable<Symbol> FilterByOpenInterest(IReadOnlyDictionary<Symbol, SecurityExchangeHours> contracts)
+        {
+            var filtered = _chainContractsLookupLimit.HasValue ? contracts.Keys.OrderBy(x => x.ID.Date).Take(_chainContractsLookupLimit.Value) : contracts.Keys;
+            var openInterest = filtered.GroupBy(x => contracts[x]).SelectMany(g => GetOpenInterest(g.Key, g.Select(i => i))).ToDictionary(x => x.Key, x => x.Value);
+
+            decimal o;
+            filtered = filtered.OrderByDescending(s => openInterest.TryGetValue(s, out o) ? o : -1).ThenBy(x => x.ID.Date);
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:79]: Maybe should orderBy directly on the resulting `openInterest`? else if the symbol is not in `openInterest` it will use -1 and it might return that symbol.
What do you think about adding something like:
```
if (openInterest.Count == 0)
{
    _algorithm.Error("OpenInterestFutureUniverseSelectionModel.FilterByOpenInterest(): failed to get historical open interest, will select no symbol");
    return Enumerable.Empty<Symbol>();
}
```

File: Algorithm.CSharp/OpenInterestFuturesRegressionAlgorithm.cs (hunk 12 above)
Comment (michael-sena) [reply Algorithm.CSharp/OpenInterestFuturesRegressionAlgorithm.cs:90]: Ah okay, noted.

File: Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs (hunk 13 above)
Comment (michael-sena) [reply Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs:79]: Okay, I've adopted a form of this.  Let me know if you're okay with that, and resolve if so.

**Comment by michael-sena** [comment]
Hazzah!
